homeassistant/components/dlna_dmr/media_player.py

```diff
@@ -24,209 +25,301 @@
+        """Handle notification from SSDP of device state change."""
+        _LOGGER.debug(
+            "SSDP notification of device %s at %s",
+            info[ssdp.ATTR_SSDP_USN],
+            info[ssdp.ATTR_SSDP_LOCATION],
+        )
 
-    def __init__(self, dmr_device, name=None):
-        """Initialize DLNA DMR device."""
-        self._device = dmr_device
-        self._name = name
+        if not self._device:
+            location = info[ssdp.ATTR_SSDP_LOCATION]
+            self.hass.async_run_job(self._device_connect, location)
+
+    async def async_config_update_listener(
+        self, hass: HomeAssistant, entry: config_entries.ConfigEntry
+    ) -> None:
+        """Handle options update by modifying self in-place."""
+        del hass  # Unused
+        _LOGGER.debug(
+            "Updating: %s with data=%s and options=%s",
+            self.name,
+            entry.data,
+            entry.options,
+        )
+        self.location = entry.data[CONF_URL]
+        self.poll_availability = entry.options.get(CONF_POLL_AVAILABILITY, False)
 
-        self._available = False
-        self._subscription_renew_time = None
+        new_event_addr = EventListenAddr(
+            entry.options.get(CONF_LISTEN_PORT) or 0,
+            entry.options.get(CONF_CALLBACK_URL_OVERRIDE),
+        )
+        if new_event_addr != self._event_addr:
+            self._event_addr = new_event_addr
+            # Changes to eventing requires a device reconnect for it to update correctly
+            await self._device_disconnect()
+            try:
+                await self._device_connect(self.location)
+            except UpnpError as err:
+                _LOGGER.debug("Couldn't (re)connect: %s", err)
+
+    async def _device_connect(self, location: str) -> None:
+        """Connect to the device now that it's available."""
+        _LOGGER.debug("Connecting to device at %s", location)
+
+        async with self._device_lock:
+            if self._device:
+                _LOGGER.debug("Trying to connect when device already connected")
+                return
+
+            domain_data = get_domain_data(self.hass)
+
+            # Connect to the base UPNP device
+            upnp_device = await domain_data.upnp_factory.async_create_device(location)
+
+            # Create/get event handler that is reachable by the device
+            event_handler = await domain_data.async_get_event_notifier(
+                self._event_addr, self.hass.loop
+            )
+
+            # Create profile wrapper
+            self._device = DmrDevice(upnp_device, event_handler)
+
+            # Set hass device information now that it's known.
+            self._attr_device_info = {
+                "name": self._device.name,
+                # Connection is based on the root device's UDN, which is
+                # currently equivalent to our UDN (embedded devices aren't
+                # supported by async_upnp_client)
```

Comment:
The one device I've come across with an embedded (i.e., not at the root level of the UPnP description XML) DMR device is Sonos speakers. Since they're already covered by the `sonos` integration, I don't think there's a pressing need to support it. But if there ever is another device like that, I have thoughts about how to support it.

Reply:
Feel free to create an issue in the `async_upnp_client` repo with your ideas. No need to implement it directly, saving the thoughts is already helpful.

Reply:
Done: https://github.com/StevenLooman/async_upnp_client/issues/80

Reply:
Thanks!